File: 卡码/代码随想录/Python/图论/bfs/LeetCode_0841.py
#841. 钥匙和房间
from typing import *
from collections import deque

class Solution:
  def canVisitAllRooms(self, rooms: List[List[int]]) -> bool:
    def bfs()->bool:
      visited:List[bool]=[False for _ in range(len(rooms))]
      q:deque[int]=deque()
      q.append(0)
      visited[0]=True
      while len(q)>0:
        k:int=q.popleft()
        keys:List[int]=rooms[k]
        for key in keys:
          if not visited[key]:
            visited[key]=True
            q.append(key)
      return all(visited)
    return bfs()

File: 卡码/代码随想录/Go/图论/bfs/LeetCode_0841.go
// 841. 钥匙和房间
package main

func canVisitAllRooms(rooms [][]int) bool {
	n := len(rooms)
	var bfs func() bool
	bfs = func() bool {
		visited := make([]bool, n)
		for i := 0; i < n; i++ {
			visited[i] = false
		}
		visited[0] = true
		q := []int{0}
		for len(q) > 0 {
			k := q[0]
			q = q[1:]
			keys := rooms[k]
			for _, key := range keys {
				if !visited[key] {
					visited[key] = true
					q = append(q, key)
				}
			}
		}
		for _, visit := range visited {
			if !visit {
				return false
			}
		}
		return true
	}
	return bfs()
}
